In [7]:
# DS4400 HW 3
# Eunchae Hong
# Problem 4: Cross Validation

import numpy as np
import pandas as pd
from sklearn.linear_model import LogisticRegression
from sklearn.discriminant_analysis import LinearDiscriminantAnalysis
from sklearn.metrics import (
    accuracy_score
)

# load the spam database in
spam_data = pd.read_csv("spambase/spambase.data", header = None)
X = spam_data.iloc[:, :-1].values
y = spam_data.iloc[:, -1].values

In [8]:
'''
    Part 1
    Implement k-fold cross-validation (CV) for training a model. The CV algorithm consists of the following steps:
        1. Divide the entire data into k partitions of equal size.
        2. Run k experiments. In each experiment, train on k-1 partitions and test on the validation set (partition i).
        3. Record the validation error for each experiment.
        4. Compute and print the average validation error across all $k$ experiments.
'''

# implement the k-fold cross validation
def k_fold_cross_val(model, X, y, k):
    N = len(X)
    fold_size = N // k

    indicies = np.arange(N)
    np.random.seed(5)
    np.random.shuffle(indicies)

    errors = []

    for i in range(k):
        validation_start = i * fold_size
        validation_end = validation_start + fold_size
        validation_indices = indicies[validation_start:validation_end]

        # set everything else to be training 
        training_indicies = np.concatenate([indicies[:validation_start], indicies[validation_end:]])
        X_train_fold = X[training_indicies]
        y_train_fold = y[training_indicies]
        X_val_fold = X[validation_indices]
        y_val_fold = y[validation_indices]


        # train and test the models
        model.fit(X_train_fold, y_train_fold)
        y_pred = model.predict(X_val_fold)

        # get the validation error
        error = 1 - accuracy_score(y_val_fold, y_pred)
        errors.append(error)
        print(f"Fold {i + 1}: Validation Error is {round(error, 3)}")

    # compute the average validation error
    avg_error = np.mean(errors)
    return avg_error

In [9]:
'''
    Part 2
    Run the CV experiment for logistic regression and LDA for k in {5,10}.
    You can use a package for training the logistic regression and LDA models.
    Print for each model the average validation error for each value of k.
'''
k_vals = [5, 10]

# train the Logistic Regression and LDA
logit_model = LogisticRegression(max_iter = 10000, solver = "lbfgs", random_state = 5)
lda_model = LinearDiscriminantAnalysis()

for val in k_vals:
    print(f"Logistic Regression (k = {val})")
    logit_avg_error = k_fold_cross_val(logit_model, X, y, val)
    print(f"Average Validation Error: {round(logit_avg_error, 3)}")

    print(f"LDA (k = {val})")
    lda_avg_error = k_fold_cross_val(lda_model, X, y, val)
    print(f"Average Validation Error: {round(lda_avg_error, 3)}")

Logistic Regression (k = 5)
Fold 1: Validation Error is 0.075
Fold 2: Validation Error is 0.083
Fold 3: Validation Error is 0.058
Fold 4: Validation Error is 0.083
Fold 5: Validation Error is 0.073
Average Validation Error: 0.074
LDA (k = 5)
Fold 1: Validation Error is 0.101
Fold 2: Validation Error is 0.136
Fold 3: Validation Error is 0.091
Fold 4: Validation Error is 0.128
Fold 5: Validation Error is 0.107
Average Validation Error: 0.113
Logistic Regression (k = 10)
Fold 1: Validation Error is 0.074
Fold 2: Validation Error is 0.067
Fold 3: Validation Error is 0.074
Fold 4: Validation Error is 0.096
Fold 5: Validation Error is 0.05
Fold 6: Validation Error is 0.059
Fold 7: Validation Error is 0.067
Fold 8: Validation Error is 0.096
Fold 9: Validation Error is 0.067
Fold 10: Validation Error is 0.07
Average Validation Error: 0.072
LDA (k = 10)
Fold 1: Validation Error is 0.1
Fold 2: Validation Error is 0.1
Fold 3: Validation Error is 0.124
Fold 4: Validation Error is 0.154
Fold 5: Val

***Part 3 Questions: Which model performs better? Compare the results.***

The **Logistic Regression** Model performs better out of the two as it consistently has the lower average validation error for k = 5 and k = 10. For k = 5 its average validation error is 0.074 and for k = 10 its average validation error is 0.072, while LDA has average validation error of 0.113 for both k values. Additionally, increasing k from 5 to 10 gives a better estimate for validation error, since each fold will use more of the training data. It also looks like the LDA model has more variability within the folds as the error values are more spread out. 